File: subsets.py
from typing import List


class Solution:
    def subsets(self, nums: List[int]) -> List[List[int]]:
        res = []
        subset = []
        def subsets(start:int):
            if start >= len(nums):
                res.append(subset.copy())
                return

            subset.append(nums[start])
            subsets(start + 1)
            subset.pop()
            subsets(start + 1)
        subsets(0)
        return res

s = Solution()
print(s.subsets([1,2,3]))

File: kotlin/src/main/kotlin/neetcode_roadmap/backtracking/Subsets.kt
package org.example.neetcode_roadmap.backtracking

fun subsets(nums: IntArray): List<List<Int>> {
    val res = mutableListOf<List<Int>>()
    val subset = mutableListOf<Int>()

    fun backtracking(i: Int = 0) {
        if (i >= nums.size) {
            res.add(subset.toList())
            return
        }
        subset.add(nums[i])
        backtracking(i + 1)
        subset.removeLast()
        backtracking(i + 1)
    }
    backtracking()
    return res
}

fun main() {
    val output = subsets(nums = intArrayOf(1, 2, 3))
    println(output)
}